core/src/com/unciv/ui/worldscreen/unit/UnitActions.kt

```diff
@@ -12,381 +12,397 @@ import com.unciv.models.UncivSound
-                        val relevantUnique = unit.getUniques().first { it.startsWith("Can undertake") }
-                        val influenceEarned = Regex("\\d+").find(relevantUnique)!!.value.toInt()
-                        tile.owningCity!!.civInfo.getDiplomacyManager(unit.civInfo).influence += influenceEarned
-                        unit.civInfo.addNotification("Your trade mission to [${tile.owningCity!!.civInfo}] has earned you [${goldEarned.toInt()}] gold and [$influenceEarned] influence!", null, Color.GOLD)
-                        unit.destroy()
-                    })
-        }
+    private suspend fun SequenceScope<UnitAction>.yieldWorkerActions(unit: MapUnit, tile: TileInfo,
+                                                                     worldScreen: WorldScreen,
+                                                                     unitTable: UnitTable) {
+        if (unit.hasUnique("Can build improvements on tiles"))
+            yieldAll(sequence {
+                yield(if (Constants.unitActionAutomation == unit.action)
+                    UnitAction(
+                            type = UnitActionType.StopAutomation,
+                            action = { unit.action = null })
+                else
+                    UnitAction(
+                            type = UnitActionType.Automate,
+                            action = getLambdaOrNull(unit.currentMovement > 0) {
+                                unit.action = Constants.unitActionAutomation
+                                WorkerAutomation(unit).automateWorkerAction()
+                            })
+                )
+
+                // Allow automate/unautomate when embarked, but not building improvements - see #1963
+                if (!unit.isEmbarked())
+                    yield(UnitAction(
+                            type = UnitActionType.ConstructImprovement,
+                            isCurrentAction = unit.currentTile.hasImprovementInProgress(),
+                            action = getLambdaOrNull(unit.currentMovement > 0
+                                    && !tile.isCityCenter()
+                                    && unit.civInfo.gameInfo.ruleSet.tileImprovements.values.any { tile.canBuildImprovement(it, unit.civInfo) }) {
+                                worldScreen.game.setScreen(ImprovementPickerScreen(tile) { unitTable.selectedUnit = null })
+                            }))
+            })
     }
 
-    private fun addFortifyActions(actionList: ArrayList<UnitAction>, unit: MapUnit, unitTable: UnitTable) {
-
-        val action = UnitAction(
-                type = UnitActionType.Fortify,
-                canAct = unit.currentMovement > 0,
-                uncivSound = UncivSound.Fortify,
-                action = {
-                    unit.fortify()
-                    unitTable.selectedUnit = null
+    private fun getLambdaOrNull(test: Boolean, lambda: () -> Unit): (() -> Unit)? =
```

Comment:
had to make this function to wrap lambdas or the type inference would fail, unless you know of a better way?